README.md
# WerkstukDevV
---
## Table of Contents
- [Project information](#project-information)
  - [Author](#author)
  - [About this project](#about-this-project)
- [Getting started](#getting-started)
- [How it works](#how-it-works)
  - [Achievements Table](#achievements-table)
  - [Genre Table](#genre-table)
- [Used dependencies](#used-dependencies)
- [Licenses](#licenses)

---
# Project information

---
## Author
- [Alex Menzies](https://github.com/AlexMenziesEHB)

---
## About this project

This Achievements API allows you to create and modify achivements and have a deeper look into their depending genre.

---
# Getting started

1. Clone this repository from github.
2. Navigate to the root folder and change .env.template to .env
   - This is a simple text configuration file for controlling your applications environment constants.
3. Navigate to the root folder and start up the Docker container.
    ```shell
    docker-compose build
    docker-compose up

    or

    docker-compose up --build
    ```
    For more information regarding Docker, please refer to [their documentation](https://docs.docker.com/).
4. To run tests with Jest, navigate to the /api folder.
    ```shell
    npm test
    ```
    For more information regarding Jest, please refer to [their documentation](https://jestjs.io/docs/getting-started).
5. To see the the data form the database, I recommend using TablePlus.
   You can get it [here](https://tableplus.com/).
6. To contrubute to this project make sure to read 'CONTRIBUTING.md' located in the root folder

---
# How it works

This Achievements API uses two tables to manage its content. Both tables are accessible through their respective CRUD endpoints.

---
## Achievements Table
This table is used to save all achievements. It follows the following structure:

|uuid  |achievementName|description|genreName|
|------|---------------|-----------|---------|
|*uuid*|*string*       |*string*   |*string* |

You can modify this table by using the following endpoints:

- `POST /achievement`
    - Posts an achievement to the database.
    - Requires a body with the following properties:
        ```js
        {
            achievementName: String,
            description: String,
            genreName: String
        }
        ```
- `GET /achievements`
    - Returns all achievements from the database.
- `GET /achievement/:uuid`
    - Returns a specific achievement by uuid.
- `PATCH /achievement/:uuid`
    - Updates the content of a specific achievement by uuid.
    - - Requires a body with the properties you wish to edit. **You do not have to provide all properties**.
        ```js
        {
            achievementName: "Welcome!",
            genreName: "Start"
        }
        ```
- `DELETE /achievement/:uuid`
    - Deletes a specific achievement by uuid.
    - Requires a body which includes the uuid as a string:
        ```js
        {
            uuid: "ab382560-580f-11ec-b011-51878d80a9a9"
        }
        ```

---
## Genre Table
This table is used to save all genres from achievements database. It follows the following structure:

|uuid  |genreName|
|------|---------|
|*uuid*|*string* |

You can modify this table by using the following endpoints:

- `POST /genre`
    - Posts an genre to the database.
    - Requires a body with the following properties:
        ```js
        {
            genreName: String
        }
        ```
- `GET /genres`
    - Returns all genres from the database.
- `GET /genre/:uuid`
    - Returns a specific genre by uuid.
- `PATCH /genre/:uuid`
    - Updates the content of a specific genre by uuid.
    - - Requires a body with the properties you wish to edit.
        ```js
        {
            genreName: "Abs"
        }
        ```
- `DELETE /genre/:uuid`
    - Deletes a specific genre by uuid.
    - Requires a body which includes the uuid as a string:
        ```js
        {
            uuid: "ab382560-580f-11ec-b011-51878d80a9a9"
        }
        ```

---
# Used dependencies
* Git flow
* Docker

- body-parser: "^1.19.0"
- express: "^4.17.1"
- http: "^0.0.1-security"
- jest: "^27.3.1"
- knex: "^0.95.13"
- nodemon: "^2.0.14"
- pg: "^8.7.1"
- puppeteer: "^12.0.1"
- sort-object-keys: "^1.1.3"
- supertest: "^6.1.6"
- uuid: "^8.3.2

---

# Licenses
Licenses for this project can be found in the root folder of this project.
Make sure to look at them and reas them before starting.
- MIT License
- Contributing
- Code of Conduct

---

CHANGELOG.md
# Changelog

---

## 1.0.0

Fixes:
- Added new database migrations

Features:
- Added database

---

### 1.0.1

Features:
- Added CRUD capabilities
- Addedintegration tests

---

### 1.0.2

Fixes:
- removed unused code

---

### 1.0.3

Features:
- Added end to end tests

---

### 1.0.4

Refactor:
- Updated test to test-ci for git actions
- updated to new dev environment

---

### 1.0.5

Features:
- Added licenses
- Added new Readme lines

---

api/__test__/helpers.test.js
const Helpers = require('../src/utils/helpers.js')

describe('stringlength tester', () => {
  test('if string has been given', () => {
    expect(Helpers.checkTitleLength()).toBeFalsy()
    expect(Helpers.checkTitleLength(101)).toBeFalsy()
    expect(Helpers.checkTitleLength([])).toBeFalsy()
  })
  test('if string length is not too much', () => {
    expect(Helpers.checkTitleLength("Hello world").length).toBeLessThan(101)
    expect(Helpers.checkTitleLength("Hello world")).toBeTruthy()
  })
  test('string starts with a capital', () => {
    expect(Helpers.checkTitleLength("hello world")).toBeFalsy()
    expect(Helpers.checkTitleLength("Hello world")).toBe("Hello world")
  })
})

describe("checkParameters test", () => {
    test("check if function returns something", () => {
      expect(Helpers.checkParameters()).not.toBeUndefined();
    })
    test("check if function returns 400 when no parameters passed", () => {
      expect(Helpers.checkParameters()).toBe(400);
    })
    test("check if function returns 400 when parameters are wrong types", () => {
      expect(Helpers.checkParameters("hey", 0, {})).toBe(400);
    })
    test("check if function returns 400 when parameters are wrong types", () => {
      expect(Helpers.checkParameters(undefined)).toBe(400);
    })
    test("check if function returns 400 when parameters are wrong types", () => {
      expect(Helpers.checkParameters(null)).toBe(400);
    })
    test("check if function returns 400 when parameters are wrong types", () => {
      expect(Helpers.checkParameters([])).toBe(400);
    })
  })

describe("generateUUID test", () => {
  test("check if generateUUID() generates something", () => {
    expect(Helpers.generateUUID()).not.toBeUndefined()
  });
  test("check if generated is UUID", () => {
    expect(Helpers.generateUUID()).toMatch(
      /\b[0-9a-f]{8}\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\b[0-9a-f]{12}\b/
    );
  });
});

api/__test__/integration.test.js
const supertest = require("supertest")
const app = require("../src/server.js")
const request = supertest(app)

/*=================== Achievements Endpoints =====================*/
describe('CRUD endpoint achievements tests', () => {
    let uuid;
    describe("GET /achievements", () => {
        it("responds with 200 and returns all achievements", async () => {
            try {
                const response = await request.get("/achievements");
                expect(response.status).toBe(200);
                expect(typeof response.body).toBe("object");
                done()
            } catch (error) {}
        });
    });

    describe("Create an achievement", () => {
        it("responds with 200 if achievement is added to db", async () => {
            try {
                const response = await request
                    .post("/achievement")
                    .send({
                        achievementName: 'New achievement',
                        description: 'This is a description of a new achievement',
                        genreName: 'Start'
                    });
                expect(response.status).toBe(200);
                uuid = response.body.uuid;
                done();
            } catch (error) {}
        });
    })

    describe("Find an achievement", () => {
        it("finds the recently added achievement in db", async () => {
            try {
                const response = await request.get("/achievement/" + uuid);
                expect(response.status).toBe(200);
                expect(typeof response.body).toBe("object");
                done();
            } catch (error) {}
        });
    })

    describe("Update an achievement", () => {
        it("responds with 200 if achievement is updated", async () => {
            try {
                const response = await request
                    .patch("/achievement/" + uuid)
                    .send({
                        description: "this is an updated description"
                    });
                expect(response.status).toBe(200);
                done();
            } catch (error) {}
        })
    })

    describe("DELETE /achievement", () => {
        it("responds with 400 if no uuid is provided", async () => {
            try {
                const response = await request.delete("/achievement").send({});
                expect(response.status).toBe(400);
                done();
            } catch (error) {}
        });
    });

});

/*=================== Genres Endpoints =====================*/

describe('CRUD endpoint genres tests', () => {
    let uuid;
    describe("GET /genres", () => {
        it("responds with 200 and returns all genres", async () => {
            try {
                const response = await request.get("/genres");
                expect(response.status).toBe(200);
                expect(typeof response.body).toBe("object");
                done()
            } catch (error) {}
        });
    });

    describe("Create a genre", () => {
        it("responds with 200 if genre is added to db", async () => {
            try {
                const response = await request
                    .post("/genre")
                    .send({
                        genreName: 'New genre'
                    });
                expect(response.status).toBe(200);
                uuid = response.body.uuid;
                done();
            } catch (error) {}
        });
    })

    describe("Find a genre", () => {
        it("finds the recently added genre in db", async () => {
            try {
                const response = await request.get("/genre/" + uuid);
                expect(response.status).toBe(200);
                expect(typeof response.body).toBe("object");
                done();
            } catch (error) {}
        });
    })

    describe("Update a genre", () => {
        it("responds with 200 if genre is updated", async () => {
            try {
                const response = await request
                    .patch("/genre/" + uuid)
                    .send({
                        genreName: "updated genre"
                    });
                expect(response.status).toBe(200);
                done();
            } catch (error) {}
        })
    })
});

api/__test__/algemenetest.test.js
const request = require("supertest");

const app = require("../src/index");

let elementId;


describe("GET / ", () => {
    test("It should respond with an array of achievements", async () => {
        const response = await request(app).get("/");
        expect(response.body.data).toEqual([{
                "description": "Succesfully signed up",
                "id": 1,
                "name": "Welcome",
                "points": "200",
            },
            {
                "description": "Succesfully tried your first workout",
                "id": 2,
                "name": "First try",
                "points": "350",
            },
            {
                "description": "Let a friend join you with your journey",
                "id": 3,
                "name": "Refer a friend",
                "points": "100",
            },
        ]);
        expect(response.statusCode).toBe(200);
    });
});


describe("Test to get the data in the database", () => {
    test("GET /", (done) => {
        request(app)
            .get("/")
            .expect("Content-Type", /json/)
            .expect(200)
            .expect((res) => {
                res.body.data.length = 1;
                res.body.data[0].name = "test achievement";
            })
            .end((err, res) => {
                if (err) return done(err);
                return done();
            });
    });
});

describe("Test to insert data into the database.", () => {
    test("POST /send", (done) => {
        request(app)
            .post("/send")
            .expect("Content-Type", /json/)
            .send({
                name: "new achievement",
            })
            .expect(201)
            .expect((res) => {
                res.body.data.length = 2;
                res.body.data[0].name = "test achievement";
                res.body.data[1].name = "new achievement";
            })
            .end((err, res) => {
                if (err) return done(err);
                elementId = res.body.data[1].id;
                return done();
            });
    });
});

describe("Test to update an element that was inserted in the database", () => {
    test("PUT /update/:id", (done) => {
        request(app)
            .put(`/update/${elementId}`)
            .expect("Content-Type", /json/)
            .send({
                name: "updated achievement",
            })
            .expect(200)
            .expect((res) => {
                res.body.data.length = 2;
                res.body.data[0].name = "test achievement";
                res.body.data[1].id = elementId;
                res.body.data[1].name = "updated achievement";
            })
            .end((err, res) => {
                if (err) return done(err);
                return done();
            });
    });
});

describe("Test to delete an element that was inserted in the database", () => {
    test("DELETE /destroy/:id", (done) => {
        request(app)
            .delete(`/destroy/${elementId}`)
            .expect("Content-Type", /json/)
            .expect(200)
            .expect((res) => {
                res.body.data.length = 1;
                res.body.data[0].name = "test achievement";
            })
            .end((err, res) => {
                if (err) return done(err);
                return done();
            });
    });
});

api/src/utils/seed.js
const pg = require("knex")({
    client: "pg",
    version: "9.6",
    searchPath: ["knex", "public"],
    connection: process.env.PG_CONNECTION_STRING ?
        process.env.PG_CONNECTION_STRING : "postgres://example:example@localhost:5432/test",
});


const SEED = {
    async initialiseData() {
        try {
            await pg.table('achievementTable').delete()
            await pg.table('genreTable').delete()

            await pg.table('achievementTable').insert([{
                id: 1,
                uuid: "d07d0d20-59e2-11ec-84da-69067e8d8904",
                achievementName: "Welcome!",
                description: "Welcome to the club!",
                genreName: "Start",
            }, {
                id: 2,
                uuid: "d07d0d20-59e2-11ec-84da-69067e8d8924",
                achievementName: "Firts start!",
                description: "Completed first workout",
                genreName: "Start",
            }]);
            await pg.table('genreTable').insert([{
                id: 1,
                uuid: "d07d0d20-59e2-11ec-84da-69067e8d8904",
                genreName: "Start",
            }, {
                id: 2,
                uuid: "d07d0d20-59e2-11ec-84da-69067e8d8924",
                genreName: "Start",
            }]);

        } catch (err) {
            console.log(err);
        }
    }
}

module.exports = SEED;

api/src/utils/helpers.js
const {v1: uuidv1 } = require('uuid');
const sortObject = require('sort-object-keys');

const Helpers = {
  generateUUID: () => {
     const uuid = uuidv1();
     return uuid;
  },

  checkTitleLength: (title) => {
    if (typeof title !== 'string' || title == null || title.length > 50 || title.length == '' || title[0] !== title[0].toUpperCase()) {
      return false;
    }
    return title;
  },

  checkParameters: (expectedParameters, givenParameters, allowOtherParameters) => {
    if(
      typeof expectedParameters == "object" &&
      typeof givenParameters == "object" &&
      typeof allowOtherParameters == "boolean"
    ) {
      let sortedExpectedParams = sortObject(expectedParameters);
      let sortedGivenParams = sortObject(givenParameters);
      let matchingParameters = 0;
      for(const requiredParam in sortedExpectedParams){
        if(
          requiredParam in sortedGivenParams &&
          typeof sortedGivenParams[requiredParam] == expectedParameters[requiredParam]
        ) {
          matchingParameters++;
        }
      }
      if(Object.keys(sortedGivenParams).length > matchingParameters && allowOtherParameters == false){
        return false;
      } else if(Object.keys(sortedGivenParams).length > matchingParameters && allowOtherParameters == true){
        return true;
      } else if(matchingParameters < Object.keys(sortedExpectedParams).length){
        return false;
      } else if(matchingParameters == Object.keys(sortedExpectedParams).length){
        return true;
      } else{
        return false;
      }
    }
    return 400;
  }
}

module.exports = Helpers

api/src/utils/database.js
const PG = require("../utils/knex.js")

const Database = {
    async initialiseTables() {
        try {
            /**
             * Create table achievemenTable
             * @param: none
             * @returns: created table achievements
             */
            await PG.schema.hasTable('achievementTable').then(async (exists) => {
                if (!exists) {
                    await PG.schema
                        .createTable('achievementTable', (table) => {
                            table.increments();
                            table.uuid('uuid');
                            table.string('achievementName');
                            table.string('description');
                            table.string('genreName');
                            table.timestamps(true, true);
                        })
                        .then(async () => {
                            console.log('created achievement table');
                        })
                }
            });
            /**
             * Create table genreTable
             * @param: none
             * @returns: created table genres
             */
            await PG.schema.hasTable('genreTable').then(async (exists) => {
                if (!exists) {
                    await PG.schema
                        .createTable('genreTable', (table) => {
                            table.increments();
                            table.uuid('uuid');
                            table.string('genreName');
                            table.timestamps(true, true);
                        })
                        .then(() => PG('achievementTable').select('achievementTable.uuid', 'achievementTable.genreName'))
                        .then((rows) => PG('genreTable').insert(rows))
                    /*.then(async () => {
                        console.log('created a genre table');
                    });*/
                }
            });
        } catch (err) {
            console.log(err);
            // process.exit(1);
        }
    }
}

module.exports = Database, PG;

api/src/server.js
const express = require('express')
const bodyParser = require('body-parser')
const http = require('http')
const Database = require("./utils/database.js");
const Helpers = require('./utils/helpers.js')

const PG = require('../src/utils/knex.js')

const app = express()
http.Server(app)


app.use(bodyParser.json())
app.use(
    bodyParser.urlencoded({
        extended: true
    })
)

/**
 * GET /test
 * @param: none
 * @returns: Testendpoint
 */
app.get("/test", (req, res) => {
    //res.status(200).send();
    res.send("hello world")
})

/*=================== Achievements Endpoints =====================*/
/**
 * GET /achievements
 * @param: none
 * @returns: Get all achievements from database
 */
app.get("/achievements", async (req, res) => {
    const result = await PG
        .select(["uuid", "achievementName", "description", "genreName", "created_at"])
        .from("achievementTable")
    res.status(200).send(result);
})

/**
 * POST /achievement
 * @param: {uuid, achievementName, description, genreName}
 * @returns: Returns uuid of the added achievement
 */
app.post("/achievement", (req, res) => {
    let uuid = Helpers.generateUUID();
    let expectedParams = {
        uuid: 'string',
        achievementName: 'string',
        description: 'string',
        genreName: 'string',
    }
    if (Helpers.checkParameters(expectedParams, {
            uuid: uuid,
            achievementName: req.body.achievementName,
            description: req.body.description,
            genreName: req.body.genreName,
            created_at: new Date()
        }, true)) {
        PG.insert({
                uuid: uuid,
                achievementName: req.body.achievementName,
                description: req.body.description,
                genreName: req.body.genreName,
                created_at: new Date(),
            })
            .into("achievementTable")
            .then(() => {
                res.json({
                    uuid: uuid
                })
            })
    } else {
        res.status(400).send("Not the correct parameters");
    }
})

/**
 * GET /achievement/uuid
 * @param: uuid
 * @returns: Get One achievement from database by uuid
 */
app.get('/achievement/:uuid', async (req, res) => {
    const result = await PG
        .select(["uuid", "achievementName", "description", "genreName", "created_at"])
        .from('achievementTable')
        .where({
            uuid: req.params.uuid
        })
    res.json({
        res: result
    })
})

/**
 * PATCH /achievement/:uuid
 * @param: Send object with the properties that need to be updated
 * @returns: Returns status code 200
 */
app.patch("/achievement/:uuid", (req, res) => {
    PG('achievementTable')
        .where({
            uuid: req.params.uuid
        })
        .update(req.body)
        .then(() => {
            res.sendStatus(200)
        })
})

/**
 * DELETE /achievement
 * @param: {uuid}
 * @returns: status code 200
 */
app.delete("/achievement", (req, res) => {
    if (Helpers.checkParameters({
            uuid: 'string'
        }, {
            uuid: req.body.uuid
        }, false)) {
        PG('achievementTable')
            .where({
                uuid: req.body.uuid
            })
            .del()
            .then(() => {
                res.sendStatus(200);
            })
    } else {
        res.status(400).send("Invalid parameters");
    }
})

/*=================== Genre Endpoints =====================*/

/**
 * GET /genres
 * @param: none
 * @returns: Get all genres from database
 */
app.get('/genres', async (req, res) => {
    const result = await PG
        .select(['uuid', 'genreName', 'created_at'])
        .from('genreTable')
    res.status(200).send(result);
})

/**
 * GET /genre/uuid
 * @param: uuid
 * @returns: Get One genre from database by uuid
 */
app.get("/genre/:uuid", async (req, res) => {
    if (Helpers.checkParameters({
            uuid: 'string'
        }, {
            uuid: req.params.uuid
        }, false)) {
        const result = await PG
            .select(["uuid", "genreName", "created_at"])
            .from("genreTable")
            .where({
                uuid: req.params.uuid
            });
        if (result.length == 0) {
            res.status(404).send("genre not found");
        } else {
            res.status(200).send(result[0]);
        }
    } else {
        res.status(400).send("Invalid parameters");
    }
});

/**
 * POST /genre
 * @param: {uuid, genreName, created_at}
 * @returns: Returns uuid of the added genre
 */
app.post("/genre", (req, res) => {
    let uuid = Helpers.generateUUID()
    let expectedParams = {
        uuid: 'string',
        genreName: 'string',
    }
    if (Helpers.checkParameters(expectedParams, {
            uuid: uuid,
            genreName: req.body.genreName,
            created_at: new Date()
        }, true)) {
        PG.insert({
                uuid: uuid,
                genreName: req.body.genreName,
                created_at: new Date(),
            })
            .into("genreTable")
            .then(() => {
                res.json({
                    uuid: uuid
                })
            })
    } else {
        res.status(400).send("Not the correct parameters");
    }

})

/**
 * PATCH /genre/:uuid
 * @param: Send object with the properties that need to be updated
 * @returns: Returns status code 200
 */
app.patch("/genre/:uuid", async (req, res) => {
    PG('genreTable')
        .where({
            uuid: req.params.uuid
        })
        .update(req.body)
        .then(() => {
            res.sendStatus(200)
        })
})

/**
 * DELETE /genre
 * @param: {uuid}
 * @returns: status code 200
 */
app.delete("/genre", (req, res) => {
    PG('genreTable')
        .where({
            uuid: req.body.uuid
        })
        .del()
        .then(() => {
            res.sendStatus(200);
        })
})

Database.initialiseTables();

module.exports = app;